includes/Hooks/FooterHook.php

```diff
@@ -0,0 +1,228 @@
+        
+        // This approach of pageID also works
+        // $pageId = $title ? $title->getArticleID() : 0;
+        // $page = $pageId ? $pageRegistry->getPageByWikiPageId( (int)$pageId ) : null;
+
+        $page = $pageRegistry->getPageByFinalTitle( mb_strtolower( $title->getPrefixedText() ) );
+        if ( !$page ) {
+            return true;
+        }
+        if ( !self::shouldDisplayFooter( $title, $skin ) ) {
+            return true;
+        }
+
+        $skin->getOutput()->addModuleStyles( 'ext.LabkiPackManager.importFooter' );
+
+        $footerHtml = self::buildFooterHtml( $page );
+        if ( $footerHtml === null ) {
+            return true;
+        }
+
+        $data .= $footerHtml;
+        return true;
+    }
+
+    /**
+     * Determine whether the footer should be shown for the current request.
+     *
+     * @param Title|null $title
+     * @param Skin $skin
+     *
+     * @return bool
+     */
+    private static function shouldDisplayFooter( ?Title $title, Skin $skin ): bool {
+        if ( !$title || !$title->canExist() ) {
+            return false;
+        }
+
+        // Only show if the config is enabled
+        $config = MediaWikiServices::getInstance()->getMainConfig();
+        if ( !$config->get( 'LabkiShowImportFooter' ) ) {
+            return false;
+        }
+
+        // Only show on standard view action to avoid cluttering edit/history pages.
+        $request = $skin->getRequest();
+        $action = $request->getVal( 'action', 'view' );
+        if ( $action !== 'view' && $action !== 'purge' ) {
+            return false;
+        }
+
+        return true;
+    }
+
+
+    /**
+     * Build the HTML snippet describing the import metadata for a page.
+     *
+     * @param Page $page Page record fetched from the registry
+     *
+     * @return string|null HTML to append or null if any dependency data is missing
+     */
+    private static function buildFooterHtml( Page $page ): ?string {
+        $services = MediaWikiServices::getInstance();
+
+        /** @var LabkiPackRegistry $packRegistry */
+        $packRegistry = $services->getService( 'LabkiPackRegistry' );
+        $pack = $packRegistry->getPack( $page->packId() );
+        if ( !$pack ) {
+            return null;
+        }
+
+        /** @var LabkiRefRegistry $refRegistry */
+        $refRegistry = $services->getService( 'LabkiRefRegistry' );
+        $ref = $refRegistry->getRefById( $pack->contentRefId() );
+        if ( !$ref ) {
+            return null;
+        }
+
+        /** @var LabkiRepoRegistry $repoRegistry */
+        $repoRegistry = $services->getService( 'LabkiRepoRegistry' );
+        $repo = $repoRegistry->getRepo( $ref->repoId() );
+        if ( !$repo ) {
+            return null;
+        }
+
+        $packName = $pack->name();
+        $repoUrl = $repo->url();
+        $refLabel = $ref->refName() ?? $ref->sourceRef();
```

Comment:
we have `refName` and `sourceRef` and they appear to be used interchangeably in ways that seem pretty error prone, especially given all the fallbacks in this extension. 

it actually doesn't look like `refName` *can* ever be set, though many places in the code attempt to access it. `sourceRef` is not nullable and `refName` is. seems like we should only access `sourceRef` and remove `refName`

Reply:
or wait? maybe that's incorrect? it seems like here and in the pack manager, rather than displaying the ref, we display the repo title from the manifest. that's confusing since one would probably expect the repo title to be the same in all refs for a given repository (still unclear why it's possible to have multiple refs for a single repository at all). it's impossible from the user interface to actually tell what ref something comes from.